pkg/apis/core/annotation_key_constants.go

```diff
@@ -82,4 +82,10 @@ const (
 	//
 	// Not all cloud providers support this annotation, though AWS & GCE do.
 	AnnotationLoadBalancerSourceRangesKey = "service.beta.kubernetes.io/load-balancer-source-ranges"
+
+	// EndpointsLastTriggerTime is the annotation key, set for endpoints objects, that represents the
+	// timestamp of the last change that triggered the endpoints object update.
+	// This annotation will be used to compute the in-cluster network programming SLI. See
+	// https://github.com/kubernetes/community/blob/master/sig-scalability/slos/network_programming_latency.md
+	EndpointsLastTriggerTime = "kubernetes.io/endpoints.last-trigger-time"
```

Comment:
As replied in the comment above, this won't be the generic last-update-time, so went with the second approach and changed to "endpoints.kubernetes.io/last-update-trigger-time".

Do you want me to document more what "update trigger" means? Currently the documentation says
> (it will be)... timestamp (...) of the last change that triggered the endpoints object update.

In general this will be the timestamp of any change that triggered the endpoints object update via endpoints_controller (if I understand correctly this is the only place that updates endpoints objects).
If the endpoints change is later ignored by the kube-proxy we won't export the SLI, but it doesn't hurt to always update the value of this annotation (it makes the definition easier).

Do you think that adding your example (with T0, T1, T2) to the documentation will make things clear?

Reply:
Yes, I want it to be very precise, please :)

Endpoints controller manages MOST endpoints, but it is legitimate to not specify a selector and instead populate endpoints manually or through some other controller.  We need to apply this annotation equally.

Reply:
I think it is OK for SLI to cover only "auto" endpoints (services with selectors) because otherwise it covers controllers we don't own.

That doesn't mean that a user couldn't write that annotation themselves in their own controller, if they have the timestamp of the trigger - would that break the SLI?  What if the user were maliciously setting it to times that were old, so as to bias the metric?

Reply:
> I think it is OK for SLI to cover only "auto" endpoints (services with selectors) because otherwise it covers controllers we don't own.

+1

> What if the user were maliciously setting it to times that were old, so as to bias the metric?

That's unfortunate, but I think we can live with it. If users are malicious, they can do much worse things (especially in scalabilty area) than biasing the metric. So personally, I wouldn't worry that much about it.

Reply:
Also, the SLO (once it will exist) will be for vanilla Kubernetes, where we don't have to worry that much about it, I think.